lib/credentials/process_credentials.js

```diff
@@ -0,0 +1,206 @@
+        });
+      }
+      var profilesFromCreds = iniLoader.loadFrom({
+        filename: this.filename ||
+          (process.env[AWS.util.configOptInEnv] && process.env[AWS.util.sharedCredentialsFileEnv])
+      });
+      for (var i = 0, profileNames = Object.keys(profilesFromCreds); i < profileNames.length; i++) {
+        profiles[profileNames[i]] = profilesFromCreds[profileNames[i]];
+      }
+      // load after profilesFromCreds to prefer profilesFromConfig
+      for (var i = 0, profileNames = Object.keys(profilesFromConfig); i < profileNames.length; i++) {
+        profiles[profileNames[i]] = profilesFromConfig[profileNames[i]];
+      }
+      var profile = profiles[this.profile] || {};
+
+      if (Object.keys(profile).length === 0) {
+        throw AWS.util.error(
+          new Error('Profile ' + this.profile + ' not found'),
+          { code: 'ProcessCredentialsProviderFailure' }
+        );
+      }
+
+      if (profile['credential_process']) {
+        this.loadViaCredentialProcess(profile, function(err, data) {
+          if (err) {
+            callback(err, null);
+          } else {
+            self.expired = false;
+            self.accessKeyId = data.AccessKeyId;
+            self.secretAccessKey = data.SecretAccessKey;
+            self.sessionToken = data.SessionToken;
+            if (data.Expiration) {
+              self.expireTime = new Date(data.Expiration);
+            }
+            callback(null);
+          }
+        });
+      } else {
+        throw AWS.util.error(
+          new Error('Profile ' + this.profile + ' did not include credential process'),
+          { code: 'ProcessCredentialsProviderFailure' }
+        );
+      }
+      this.expired = false;
+      if (this.accessKeyId && this.secretAccessKey) {
+        callback(null);
+      }
+    } catch (err) {
+      callback(err);
+    }
+  },
+
+  /**
+  * Executes the credential_process and retrieves
+  * credentials from the output
+  * @api private
+  * @param profile [map] credentials profile
+  * @throws SharedIniFileCredentialsProviderFailure
+  */
+  loadViaCredentialProcess: function loadViaCredentialProcess(profile, callback) {
+    proc.exec(profile['credential_process'], function(err,stdOut, stdErr) {
+      try {
+        var credData = JSON.parse(stdOut);
+        if (credData.Expiration) {
+          var currentTime = AWS.util.date.getDate();
+          var expireTime = new Date(credData.Expiration);
+          if (expireTime < currentTime) {
+            err = AWS.util.error(
+              new Error('credential_process returned expired credentials'),
+              { code: 'ProcessCredentialsProviderFailure' }
+            );
+          }
+        } else if (credData.Version !== 1) {
+          err = AWS.util.error(
+            new Error('credential_process does not return Version == 1'),
+          { code: 'ProcessCredentialsProviderFailure' }
+          );
+        }
+        if (err) {
+          callback(err, null);
```

Comment:
Should all cases where `err` is not known to have already been wrapped, be wrapped with `AWS.util.error` and the code assigned?